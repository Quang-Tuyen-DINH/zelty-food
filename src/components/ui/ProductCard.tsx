import React, { useState } from 'react';
import uuid from 'react-uuid';
import { ProductCardStyled } from '../styles/ProductCard.Styled';
import { Product } from "../../shared/models/Product.model";
import { Badge } from './Badge';
import { useDispatch } from 'react-redux';
import { Button } from './Button';
import { Item } from '../../shared/models/Option.model';
import Notification from '../../features/Notification';

const ProductCard = (props: {product: Product, options?: Item[]}) => {
  const dispatch = useDispatch();
  const [chooseOption, setChooseOption] = useState<boolean>(false);
  const [selectedOption, setSelectedOption] = useState<string>("");

  const addToCartWithOptions = () => {
    setChooseOption(false);
    dispatch({ type: "ADD_PRODUCT", payload: {commandId: uuid(), productId: props.product.id, options: selectedOption, price: props.product.price} });
    Notification.notifyProduct("addProduct", props.product.name);
  }

  const addToCart = () => {
    dispatch({ type: "ADD_PRODUCT", payload: {commandId: uuid(), productId: props.product.id, price: props.product.price} });
    Notification.notifyProduct("addProduct", props.product.name);
  }

  return (
    <ProductCardStyled className="product-card">
      <header className="product-card__header">
        <img className="product-card__header__image" alt={ProductCard.name} src={props.product.image}/>
        <div className="product-card__header__button">
          {props.product.sold_out === true ?
          <Button disabled>En rupture</Button>
          :
          (props.product.available_options ?
            (chooseOption === false ?
              (props.product.available_options[0][0] === "S" ?
                <Button onClick={() => setChooseOption(true)}>Choisir la sauce</Button>
              :
                <Button onClick={() => setChooseOption(true)}>Choisir l'accompagnement</Button>
              )
            :
              <Button onClick={addToCartWithOptions}>Confirmer</Button>
            )
            :
            <Button onClick={addToCart}>Choisir</Button>
          )
          }
        </div>
      </header>
      <div className="product-card__content">
        <h3 className="product-card__content__name">{props.product.name}</h3>
        {chooseOption === false ?
          <span className="product-card__content__description">{props.product.description}</span>
        :
          <div className="product-card__content__options">
            <div onClick={() => setChooseOption(false)} className="product-card__content__options__back">
              <i className="product-card__content__options__back__arrow"></i><p>Retour</p>
            </div>
            {(props.product.available_options && props.options) &&
              props.product.available_options.map((option: string, index) => (
                <label key={`option-${index}`} className="product-card__content__options__radio">
                  <input name="radio" type="radio" value={option} onClick={() => setSelectedOption(option)}/>
                  <span className="product-card__content__options__radio__label">
                    {props.options && props.options.find(item => item.id === option) ?
                      props.options.find(item => item.id === option)?.name
                    :
                      <></>}
                    </span>
                </label>
              ))
            }
          </div>
        }
      </div>
      <div className="product-card__footer">
        {props.product.sold_out ?
          <>
            <span className="product-card__footer__price-sold-out">{(props.product.price/100).toFixed(2)} €</span>
            <Badge warning>
              Sold out
            </Badge>
          </>
        :
          <span className="product-card__footer__price">{(props.product.price/100).toFixed(2)}€</span>
        }
      </div>
    </ProductCardStyled>
  )
}

export default ProductCard