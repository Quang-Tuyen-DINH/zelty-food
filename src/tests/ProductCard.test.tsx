import { render, screen, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux'
import Adapter from "@cfaester/enzyme-adapter-react-18";
import { shallow, configure } from "enzyme";
import ProductCard from '../components/ui/ProductCard';
import Store from '../store/Index';

afterEach(cleanup);

configure({adapter: new Adapter()});

const testProduct = {
    menuId: "M02",
    image: "https://zelty-front-test-storage.s3.eu-west-1.amazonaws.com/burger-poulet.jpeg",
    description: "Bun's brioché du boulanger aux flocons d'épeautre, 2 grosses aiguillettes de poulet mariné et pané maison, cheddar mature, oignons rouge pickles, sauce mayonnaise et salade batavia",
    id: "Z1528",
    price: 900,
    sold_out: false,
    available_options: [
    "A01",
    "A02"
    ],
    name: "Burger poulet"
}

describe("Test ProductCard", () => {
  it('tests the name', () => {
  
    const product = shallow(
      <Provider store={Store}>
        <ProductCard product={testProduct} />
      </Provider>
    ).dive()
  
    expect(product).toMatchSnapshot();
  });
})