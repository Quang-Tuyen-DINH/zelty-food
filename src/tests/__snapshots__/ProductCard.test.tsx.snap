// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ProductCard tests the name 1`] = `ShallowWrapper {}`;
